### ISSUE TITLE
Upgrade to 3.5.0

### ISSUE THREAD

**Issue by Wulfsta** [nixos-rocm/nixos-rocm#41]
It looks like they've skipped 3.4.0 entirely and went right to 3.5.0? Let me know if you want help adding tests to anything in particular or trying to switch to ROCclr. I'd like to work towards getting the high level ML libraries in a working state again, if the segfaults still happen in this version.

**Comment by acowley** [comment]
I did start the 3.5 upgrade, but not pushed the WIP branch here yet. I’ll do that asap.

**Comment by Wulfsta** [comment]
Thanks!

**Comment by acowley** [comment]
I've pushed where I am [here](https://github.com/acowley/nixos-rocm/tree/3.5.x), but it's not great. I have the core pieces building: `roct`, `rocr`, `rocclr`, and `rocm-opencl-runtime`. I also have `hip-clang` building, but executables it produces segfault. My test program segfaults in `hipMalloc`, for example.

I don't know if OpenCL is working with this yet.

I used a ROCm docker image to build the test executable, and flags passed to `clang++` as well as `ldd` output agree with what I see on NixOS.

One [patch](https://github.com/acowley/nixos-rocm/blob/3.5.x/pkgs/development/libraries/rocclr/default.nix#L31) I had to make bugs me because I couldn't find a way to fix it in the CMakeLists.txt.

I don't currently have any leads on the segfault, so help is more than welcome.

**Comment by acowley** [comment]
Huzzah! I found the issue 🙌. It was another instance where they open a dynamically linked library in code, and the search paths used were not sufficient 💆‍♂️. I replaced the search in the code with the absolute path to the library in the store.

So now hip-clang can compile a test executable for me. Please do jump in to help with the libraries that sit atop it.

**Comment by acowley** [comment]
Also, if people have feelings about merging this 3.5 branch to master, please let me know. I'm fine with doing it since I think we now have working OpenCL and HIP (though I'd like to be more confident that both of those are working properly). But of course it would be a (temporary) step back in how many ROCm packages we offer, so waiting until we improve coverage is fine, too.

**Comment by Wulfsta** [comment]
> So now hip-clang can compile a test executable for me. Please do jump in to help with the libraries that sit atop it.

Sorry, I have not gotten to sit down and work on this yet. Hopefully I'll have time in the next few days! 

I'm fine with merging this, but think we should keep this issue open (or open a new one) to track what builds and what has tests, and consolidate all related issues into one. I think if we do this we can close #34 and #39, as well as include efforts towards compiling TensorFlow from source if you think that's appropriate. I think feature requests for new packages should remain separate.

**Comment by athas** [comment]
I've tried [the branch](https://github.com/acowley/nixos-rocm/tree/3.5.x), but I have a weird issue where my generated `/var/run/opengl-driver/etc/OpenCL/vendors/amdocl64.icd` contains `/nix/store/8gni2pprx8c2am93396i1q3c513xjkn9-rocm-opencl-runtime-3.5.0/lib/x86_64/libamdocl64.so`, which does not exist.  I think it should contain `/nix/store/8gni2pprx8c2am93396i1q3c513xjkn9-rocm-opencl-runtime-3.5.0/lib/libamdocl64.so`, which does exist (or `libamdocl64.so` should be put elsewhere).

**Comment by athas** [comment]
Applying this patch fixes it for me:

```
diff --git a/pkgs/development/libraries/rocm-opencl-icd.nix b/pkgs/development/libraries/rocm-opencl-icd.nix
index 13467b0..ad7e9e6 100644
--- a/pkgs/development/libraries/rocm-opencl-icd.nix
+++ b/pkgs/development/libraries/rocm-opencl-icd.nix
@@ -3,7 +3,7 @@
 stdenv.mkDerivation rec {
   version = "3.3.0";
   name = "rocm-opencl-icd";
-  src = writeText "amdocl64.icd" "${rocm-opencl-runtime}/lib/x86_64/libamdocl64.so";
+  src = writeText "amdocl64.icd" "${rocm-opencl-runtime}/lib/libamdocl64.so";
   unpackPhase = ":";
   installPhase = ''
     mkdir -p $out/etc/OpenCL/vendors
```

**Comment by acowley** [comment]
Thanks for catching that! I've pushed the fix.

**Comment by acowley** [comment]
We're on the cusp of having cachix enabled, but I jumped the gun by creating the `nixos-rocm` cachix without realizing that I don't have access to change the settings of this repository. @peter-sa can you give me permission to adjust the settings of the repository?

**Comment by acowley** [comment]
Easier solution: I have my nixos-rocm fork pushing to the nixos-rocm cachix for now.

ROCm-3.5.0 OpenCL, hip-clang, and rocblas will all be there as soon as the CI finishes with rocblas.

**Comment by acowley** [comment]
As an update now that things have cycled a couple times. Since I was able to enable the push to cachix on the `acowley/nixos-rocm` fork, the [CI time](https://github.com/acowley/nixos-rocm/actions/runs/134717427) on a minor README tweak there was 3m43s.

This compares very favorably to the 3+ hour time of a real build of that job set!

**Comment by athas** [comment]
Cool!  That sounds downright tolerable.

**Comment by acowley** [comment]
Things with CI [are looking great](https://github.com/acowley/nixos-rocm/actions/runs/148027059). We’ve got all the heavy builds going, and the Tensorflow 1 wheel works for some things, but not others (the old cnn benchmark fails for me on gfx803, it might work with Vega).

@peter-sa can you temporarily grant me access to add the cachix signing key to this repo, or would you prefer we keep development on my fork and periodically push here? That would be fine, I think, as everyone could still benefit from cachix.

I’d appreciate if anyone with Vega can try out TF1, and maybe take a stab at updating the TF2 wheel, or even the pytorch build.

**Comment by Wulfsta** [comment]
@acowley I'll try to test TensorFlow 1 and see if I can get the other ML libraries building this weekend, but no promises - I have some IRL stuff happening at the moment.

**Comment by Wulfsta** [comment]
@acowley, can you push your `3.5.x` branch to this repository?

**Comment by acowley** [comment]
@Wulfsta Can you use my fork as a separate remote? With cachix in play, we can potentially deal with CI timeouts since we'll make progress each time, but we can't do that here as it requires modifying settings that we don't have permissions for. My plan was to move development onto our forks, while using this repo for releases and perhaps as a central issue tracker.

**Comment by Wulfsta** [comment]
@acowley sure, no problem!

**Comment by Wulfsta** [comment]
After a quick test I can confirm that the TensorFlow 1 derivation works on `gfx906`. I'm still having nondeterministic failures compiling `rocfft` with parallel building disabled, haven't tracked that down. I have not had a chance to work towards TensorFlow 2 or PyTorch. I am most interested in compiling TensorFlow 2 from source, so I will probably focus my efforts there.

**Comment by Wulfsta** [comment]
@acowley I was building an attempt at TensorFlow 2 and noticed the message `1 warning generated when compiling for gfx803` appear, but I have `rocmTargets = [ "gfx906" ]` set. Any idea what package might be causing this? 

Edit: In fact, `1 warning generated when compiling for gfx900` just popped up as well. As I recall defaults are `gfx803, gfx900, gfx906`: is there a derivation that is failing to pick up (or use) `rocmTargets`?

Edit 2: ~~This might be Tensile? I haven't checked, but that is what I am currently suspicious of.~~ This was wrong, happened earlier in the build. `rccl` maybe?

**Comment by acowley** [comment]
I noticed something different with targets being set by  upstream packages. You can see that I had to set the targets in `rocrand` for the 3.5 upgrade. Perhaps you’re seeing another instance of that, or maybe I messed that up.

**Comment by acowley** [comment]
I merged the `3.5.x` branch into `master`. Letting `master` linger was starting to cause some friction with PRs.
